src/java/org/apache/cassandra/index/sai/StorageAttachedIndex.java

```diff
@@ -542,26 +543,36 @@ public RowFilter getPostIndexQueryFilter(RowFilter filter)
     }
 
     @Override
-    public Comparator<List<ByteBuffer>> getPostQueryOrdering(Restriction restriction, int columnIndex, QueryOptions options)
-    {
-        // For now, only support ANN
+    public void postQuerySort(ResultSet cqlRows, Restriction restriction, int columnIndex, QueryOptions options) {
         assert restriction instanceof SingleColumnRestriction.AnnRestriction;
 
         Preconditions.checkState(indexContext.isVector());
 
         SingleColumnRestriction.AnnRestriction annRestriction = (SingleColumnRestriction.AnnRestriction) restriction;
-        VectorSimilarityFunction function = indexContext.getIndexWriterConfig().getSimilarityFunction();
-
-        float[] target = TypeUtil.decomposeVector(indexContext, annRestriction.value(options).duplicate());
-
-        return (leftBuf, rightBuf) -> {
-            float[] left = TypeUtil.decomposeVector(indexContext, leftBuf.get(columnIndex).duplicate());
-            double scoreLeft = function.compare(left, target);
-
-            float[] right = TypeUtil.decomposeVector(indexContext, rightBuf.get(columnIndex).duplicate());
-            double scoreRight = function.compare(right, target);
-            return Double.compare(scoreRight, scoreLeft); // descending order
-        };
+        VectorSimilarityFunction similarityFunction = indexContext.getIndexWriterConfig().getSimilarityFunction();
+
+        float[] targetVector = TypeUtil.decomposeVector(indexContext, annRestriction.value(options).duplicate());
+
+        List<List<ByteBuffer>> buffRows = cqlRows.rows;
+        // Decorate-sort-undecorate to optimize sorting of vectors by their similarity scores
+        List<Pair<ByteBuffer, Double>> listPairsVectorsScores = buffRows.stream()
+                                                                        .map(row -> {
+                                                                            ByteBuffer vectorBuffer = row.get(columnIndex);
+                                                                            float[] vector = TypeUtil.decomposeVector(indexContext, vectorBuffer.duplicate());
+                                                                            Double score = (double) similarityFunction.compare(vector, targetVector);
+                                                                            return new Pair<>(vectorBuffer.duplicate(), score);
```

Comment:
I think we want to store the `row` instead of the `vectorBuffer`. That way, when we undecorate the sorted list, we just have to call `pair.left`.

Reply:
Great catch, thanks. Addressed at [f10da6c](https://github.com/datastax/cassandra/pull/710/commits/f10da6c38a9271d5aff49da93b24b5e6649e8d08).